{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Importing files\n",
    "from nupack import * # imports nupack\n",
    "import pandas as pd # imports pandas\n",
    "import numpy as np\n",
    "import matplotlib.pyplot as plt # imports plotting functions\n",
    "import re\n",
    "import random\n",
    "import csv\n",
    "import ast \n",
    "import sys"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "######################################## CHANGE ########################################\n",
    "# Additional Strands\n",
    "BindingSite = \"CAATAAGCGATGCGCCCTCGCCTGGGGGCCTAGTCCTCT\" # 5'-3'\n",
    "sBinding = \"GTTATTCGCTACG\" #3'-5' # short binding site\n",
    "Segment2 = \"CCTATGCGTGCTACCGTGAA\" #5'-3' # ye duan's primer\n",
    "yeDuanAptamer = \"AGCAGCACAGAGGTCAGATGCAATAAGCGATGCGCCCTCGCCTGGGGGCCTAGTCCTCTCCTATGCGTGCTACCGTGAA\" #5'-3'\n",
    "\n",
    "# Primers\n",
    "# Mark 10 \n",
    "fwdPrimer = \"CGTTCAATGATATAGGCTCGCATAGGAGACAC\" #5'-3'\n",
    "cFwdPrimer = \"GCAAGTTACTATATCCGAGCGTATCCTCTGTG\" #3'-5' #\"CACAGAGGATACGCTCGGATATAGTAACTTGC\"# Reverse complement #5'-3'\n",
    "rvsPrimer = \"CAAAGCTAGAAGAGCAGCACAGAGGTCAGATG\" #5'-3' \n",
    "cRvsPrimer = \"GTTTCGATCTTCTCGTCGTGTCTCCAGTCTAC\" #3'-5' #\"CATCTGACCTCTGTGCTGCTCTTCTAGCTTTG\"# Reverse complement 5'-3'\n",
    "\n",
    "# Constants\n",
    "nitrogenousBase = (\"A\",\"T\",\"C\",\"G\")\n",
    "nucleicAcid = 'DNA'\n",
    "temperature = 37                                                                    # In celcius. Can change to kelvin in my_model or simple conversion.\n",
    "ionNa = 0.100                                                                       # Nupack boundaries molar [0.05, 1.1] molar\n",
    "ionMg = 0.002                                                                       # Nupack boundaries molar [0.0, 0.2] molar\n",
    "concentration = 5e-7                                                                # In molar. Ex: 5e-7 is 0.5 uM\n",
    "sizeMax = 2                                                                         # max size for complexes\n",
    "gcMax = 0.70                                                                        # GC content max value\n",
    "gcMin = 0.30                                                                        # GC content min value\n",
    "probeLenMax = 46                                                                    # Max length of PROBE binding site\n",
    "probeLenMin = 46                                                                    # Min length of PROBE binding site\n",
    "gapLenMax = 20                                                                      # Max length of between PROBE and PRIMER binding site\n",
    "gapLenMin = 20                                                                      # Min length of between PROBE and PRIMER binding site\n",
    "n = 20000                                                                           # total number of aptamer and blocker pairs\n",
    "minGibbs = -12                                                                      # minimum accepted Gibbs Free Energy in kcal/mol\n",
    "maxGibbs = -5                                                                       # maximum accepted Gibbs Free Energy kcal/mol\n",
    "gPattern = re.compile(r'^#.{0.2}(GGG)')                                             # No consecutive G bases from the first 0 to 2 bases\n",
    "consecutiveBP = re.compile(r'(GGGGGG|CCCCC|AAAAA|TTTTT)')                           # Max mononucleotide 5 (ex: GGGGG, etc) [Note: G is maxed to 6 since pre-existing aptamer has 5 already]\n",
    "\n",
    "#  CHANGE NAMES EVERYTIME! IT WILL OVERWRITE FILES WITH THE SAME NAME\n",
    "openExistingFile_Boolean = False                                                    # Boolean to access sequences from a different file\n",
    "openExistingFile_Name = 'Mark9.csv'                                                 # Name of existing csv file\n",
    "saveProbeGap_Sequences = True                                                       # Boolean to save the probe and gap sequences generated in one run\n",
    "saveSequences_Name = \"Mark2Version2.csv\"                                            # Name of existing csv file\n",
    "saveData_PreFilters_Name = \"Analysized Aptamer and Blockers Mark2Version2.xlsx\"     # Name of excel file containing prefilter analysis on sequences \n",
    "saveData_PostFilters_Name = \"1_20Gap_46Probe_M2V2.xlsx\"                             # Name of excel file containing final sequences after filters\n",
    "\n",
    "# Physical model\n",
    "my_model = Model(material=nucleicAcid, celsius=temperature, sodium=ionNa, magnesium=ionMg) \n",
    "###########################################################################################"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Functions\n",
    "def gcContents(seq):\n",
    "    ''' gcContents finds gc fraction of the sequence \n",
    "    Calculates the total G's and C's in the string\n",
    "    Parameters:     seq                 string of oligonucleotides \n",
    "    Returns:        counter/len(seq)    decimal fraction\n",
    "    '''\n",
    "    counter = seq.count(\"G\")+seq.count(\"C\")\n",
    "    if len(seq) > 0:\n",
    "        gC = counter/len(seq)\n",
    "    return gC\n",
    "\n",
    "def cDNAorRNA(seq):\n",
    "    ''' cDNAorRNA returns the complementary sequence of input\n",
    "    Parameters:     seq                 Oligonucleotide sequence\n",
    "    Returns:        seq                 Complementary oligonucleotide sequence\n",
    "    '''\n",
    "    relationship = {'A':'T','T':'A','C':'G','G':'C'}\n",
    "    return \"\".join([relationship[base] for base in seq])\n",
    "\n",
    "def reverseComp(seq):\n",
    "    complement = {'A': 'T', 'T': 'A', 'C': 'G', 'G': 'C'}\n",
    "    return ''.join(complement[base] for base in reversed(seq))\n",
    "\n",
    "def randomSeq(base, maxLen, minLen):\n",
    "    ''' randSeq creates a random oligonucleotide sequence\n",
    "    Takes the parameter 'bases' and randomizes it\n",
    "    Parameters:     bases               bases of nucleotides\n",
    "                    max                 Maximum sequence length\n",
    "                    min                 Minimum sequence length\n",
    "    Returns:        seq                 Oligonucleotide sequence\n",
    "    '''\n",
    "    counter = range(random.randint(minLen,maxLen))\n",
    "    for i in counter:\n",
    "        seq = \"\".join([random.choice(base) for i in counter]) \n",
    "    return seq\n",
    "\n",
    "def generateSequences (bases, minGC, maxGC, minLen, maxLen, trigger, limit):\n",
    "    ''' generateSequences generates x amount of random oligonucleotide sequences\n",
    "    that satisfys user's GC content and sequence length range \n",
    "    Parameters:     bases               bases of nucleotides\n",
    "                    minGC               Minimum GC content\n",
    "                    maxGC               Maximum GC content\n",
    "                    maxLen              Maximum sequence length\n",
    "                    minLen              Minimum sequence length\n",
    "                    trigger             Determines where to store data\n",
    "                    limit               Number of sequences to generate\n",
    "    '''\n",
    "    global cProbe, probeLen, gap, gapLen\n",
    "    for i in range(limit):\n",
    "        gC = 0\n",
    "        while gC > maxGC or gC < minGC:\n",
    "            sequence = randomSeq(bases, maxLen, minLen)\n",
    "            gC = gcContents(sequence)\n",
    "        if trigger == \"Probe\":\n",
    "            cProbe.append(sequence)\n",
    "            probeLen.append(len(sequence))\n",
    "        elif trigger == \"Gap\":\n",
    "            gap.append(sequence)\n",
    "            gapLen.append(len(sequence))\n",
    "\n",
    "    if trigger == \"Probe\":\n",
    "        print(f\"Completed generating {len(cProbe)} probe region sequences\")\n",
    "    elif trigger == \"Gap\":\n",
    "        print(f\"Completed generating {len(gap)} gap region sequences\")\n",
    "\n",
    "def packagingData(data, seq, strandList, gEnergyList, mfeList):\n",
    "    ''' packagingData sorts data to specific lists \n",
    "    Parameters:     data                Unfiltered (raw) data \n",
    "                    seq                 Oligonucleotide sequence\n",
    "                    strandList          Lists to store seq\n",
    "                    gEnergyList         Lists to store complex gibbs energy\n",
    "                    mfeList             Lists to store minimum free energy\n",
    "    '''\n",
    "    # data: model, strand name, pFunc, gibbs, mfe, none, pairs\n",
    "    if seq != None:\n",
    "        strandList.append(seq)\n",
    "    gEnergyList.append(data[3])\n",
    "    mfeList.append(next((energy for structure, energy, stackE in data[8]), None))\n",
    "\n",
    "def read_mark_csv(csv_filename):\n",
    "    ''' read_mark_csv reads data saved into a csv file\n",
    "    Parameters:     csv_filename        Name of csv file\n",
    "    Returns:        gap_sequences       List of gap sequences\n",
    "                    probe_sequences     List of probe sequences\n",
    "    '''\n",
    "    gap_sequences = []\n",
    "    probe_sequences = []\n",
    "    \n",
    "    with open(csv_filename, 'rt') as csvFile:\n",
    "        reader = csv.reader(csvFile)\n",
    "        next(reader) \n",
    "        for row in reader:\n",
    "            gap_sequences.append(row[1])  \n",
    "            probe_sequences.append(row[2])\n",
    "    return gap_sequences, probe_sequences\n",
    "\n",
    "def graphingAptamerBlockerCAP(concList, index, titleName, labels, variable_1):\n",
    "    ''' graphingConcentration3Strand graphs 3 strand data \n",
    "    Parameters:     concList            List of concentrations \n",
    "                    index               Sequence Pair index\n",
    "                    titleName           Name of the title\n",
    "                    labels              Labels for the legend\n",
    "    '''\n",
    "    concList = [np.array(conc) for conc in concList]\n",
    "    a, aa, aaa, aab, aac, ab, abb, abc, ac, acb, acc,b, bb, bbb, bbc, bc, bcc, c, cc, ccc = concList\n",
    "    total = [sum(complex) for complex in zip(*concList)]\n",
    "    colors =    [\"#1f77b4\", #a blue\n",
    "                \"#ff7f0e\", #b orange\n",
    "                \"#D7D731\", #c yellow\n",
    "                \"#d62728\", #ab red\n",
    "                \"#82011F\", #abc rose\n",
    "                \"#FFBF07\", #bc \n",
    "                \"#2ca02c\", #aa green\n",
    "                \"#206E20\", #aaa dark green\n",
    "                \"#31BE94\", #aab sea green\n",
    "                \"#9BEC04\", #aac lime green\n",
    "                \"#5f8cbf\",#abb light blue \n",
    "                \"#98F5F9\",#ac  ???\n",
    "                \"#A7B1BE\", #acb grey\n",
    "                \"#DBAAE2\", #acc pink??\n",
    "                \"#9467bd\", #bb purple\n",
    "                \"#C58D85\",#bbb coral\n",
    "                \"#b48ae6\", #bbc lavendar\n",
    "                \"#3A3728\", #bcc mud\n",
    "                \"#F4F4BD\", #cc ray gold\n",
    "                \"#838307\"#ccc dark gold\n",
    "                ]\n",
    "    plt.figure(figsize=(15,5))\n",
    "    barWidth = 0.15\n",
    "    xAxis = np.arange(len(index))\n",
    "    offset1 = xAxis-barWidth\n",
    "    offset2 = xAxis\n",
    "    offset3 = xAxis+barWidth\n",
    "\n",
    "    # Neutral\n",
    "    plt.bar(offset2, a/total, barWidth, label = labels[0], color = colors[0])\n",
    "    bar2 = a/total\n",
    "    plt.bar(offset2, c/total, barWidth, label = labels[17], bottom = bar2, color = colors[2])\n",
    "    bar2 += c/total\n",
    "    \n",
    "    # Benefitial \n",
    "    plt.bar(offset1, b/total, barWidth, label = labels[11], color = colors[1])\n",
    "    bar1 = b/total\n",
    "    plt.bar(offset1, ac/total, barWidth, label = labels[8], bottom = bar1, color = colors[11])\n",
    "    bar1 += ac/total\n",
    "\n",
    "    # Detrimental\n",
    "    plt.bar(offset3, aa/total, barWidth, label = labels[1], color = colors[6])\n",
    "    bar3 = aa/total\n",
    "    plt.bar(offset3, aaa/total, barWidth, label = labels[2], bottom = bar3, color = colors[7])\n",
    "    bar3 += aaa/total\n",
    "    plt.bar(offset3, aab/total, barWidth, label = labels[3], bottom = bar3, color = colors[8])\n",
    "    bar3 += aab/total\n",
    "    plt.bar(offset3, aac/total, barWidth, label = labels[4], bottom = bar3, color = colors[9])\n",
    "    bar3 += aac/total\n",
    "    plt.bar(offset3, ab/total, barWidth, label = labels[5], bottom = bar3, color = colors[3])\n",
    "    bar3 += ab/total\n",
    "    plt.bar(offset3, abb/total, barWidth, label = labels[6], bottom = bar3, color = colors[10])\n",
    "    bar3 += abb/total\n",
    "    plt.bar(offset3, acb/total, barWidth, label = labels[9], bottom = bar3, color = colors[12])\n",
    "    bar3 += acb/total\n",
    "    plt.bar(offset3, acc/total, barWidth, label = labels[10], bottom = bar3, color = colors[13])\n",
    "    bar3 += acc/total\n",
    "    plt.bar(offset3, bb/total, barWidth, label = labels[12], bottom = bar3, color = colors[14])\n",
    "    bar3 += bb/total\n",
    "    plt.bar(offset3, bbb/total, barWidth, label = labels[13], bottom = bar3, color = colors[15])\n",
    "    bar3 += bbb/total\n",
    "    plt.bar(offset3, bbc/total, barWidth, label = labels[14], bottom = bar3, color = colors[16])\n",
    "    bar3 += bbc/total\n",
    "    plt.bar(offset3, bcc/total, barWidth, label = labels[16], bottom = bar3, color = colors[17])\n",
    "    bar3 += bcc/total\n",
    "    plt.bar(offset3, cc/total, barWidth, label = labels[18], bottom = bar3, color = colors[18])\n",
    "    bar3 += cc/total\n",
    "    plt.bar(offset3, ccc/total, barWidth, label = labels[19], bottom = bar3, color = colors[19])\n",
    "    bar3 += ccc/total\n",
    "    plt.bar(offset3, abc/total, barWidth, label = labels[7], bottom = bar3, color = colors[4])\n",
    "    bar3 += abc/total\n",
    "    plt.bar(offset3, bc/total, barWidth, label = labels[15], bottom = bar3, color = colors[5])\n",
    "    bar3 += bc/total\n",
    "\n",
    "    plt.legend(loc='right', bbox_to_anchor=(1.15,0.5))\n",
    "    plt.title(f\"{titleName}\")\n",
    "    if len(index) <= 300:\n",
    "        plt.xticks(xAxis, index, rotation = -90)\n",
    "    plt.ylabel('Probabilty')\n",
    "    plt.xlabel('Sequence Index')\n",
    "    plt.ylim(0,1)\n",
    "\n",
    "    plt.figure(figsize=(5,5))\n",
    "    barWidth = 0.15\n",
    "    xAxis = np.arange(len(index))\n",
    "    x1 = (ac/total)*100\n",
    "    y1 = (b/total)*100\n",
    "    \n",
    "    unqiueLabels = {}\n",
    "    shadingHue = ['orange', '#5f8cbf', 'Green']\n",
    "    colorGroup = ['[0.5-0.65]', '(0.65-0.8]', '(0.8-1]'] \n",
    "    legendShapes = ['o','s','^'] #circle, square, triangle\n",
    "    \n",
    "    for i in range(len(variable_1)):\n",
    "        val = variable_1[i]\n",
    "        if val <= 0.65:\n",
    "            num = 0\n",
    "        elif 0.65 < val <= 0.8:\n",
    "            num = 1\n",
    "        else:\n",
    "            num = 2\n",
    "        \n",
    "        plt.scatter(x1[i], y1[i], color=shadingHue[num], marker=legendShapes[num])\n",
    "        plt.text(x1[i], y1[i] + 0.1, str(i+1), ha='center', va='bottom') #chr(65+i) =letters; str(i+1) =numbers\n",
    "        \n",
    "        if colorGroup[num] not in unqiueLabels:\n",
    "            unqiueLabels[colorGroup[num]] = plt.Line2D([],[], color=shadingHue[num],\n",
    "                                                       marker=legendShapes[num],\n",
    "                                                       label=colorGroup[num],\n",
    "                                                       linestyle = 'None')\n",
    "           \n",
    "    plt.xlabel('A.C (%)', fontsize = 14)\n",
    "    plt.ylabel('B (%)', fontsize =14)\n",
    "    plt.legend(\n",
    "        handles=list(unqiueLabels.values()),\n",
    "        loc='upper right',\n",
    "        bbox_to_anchor=(1.35,1),\n",
    "        title = 'Ratio of A.B in\\nabsence of CAP')\n",
    "    plt.title('Effect of CAP on Aptamer-Blocker System', fontsize=14)\n",
    "\n",
    "def graphingConcentration3Strand(concList, index, titleName, labels):\n",
    "    ''' graphingConcentration3Strand graphs 3 strand data \n",
    "    Parameters:     concList            List of concentrations \n",
    "                    index               Sequence Pair index\n",
    "                    titleName           Name of the title\n",
    "                    labels              Labels for the legend\n",
    "    Returns:        total               List of the sums of concentrations present in their respective reaction\n",
    "    '''\n",
    "    concList = [np.array(conc) for conc in concList]\n",
    "    a, aa, aaa, aab, aac, ab, abb, abc, ac, acb, acc,b, bb, bbb, bbc, bc, bcc, c, cc, ccc = concList\n",
    "    total = [sum(complex) for complex in zip(*concList)]\n",
    "    colors =    [\"#E69F00\", \"#56B4E9\", \"#009E73\", \"#F0E442\", \"#0072B2\", \"#D55E00\", \"#CC79A7\", \"#999999\",\n",
    "                \"#000000\", \"#FF0000\", \"#8E44AD\", \"#1F77B4\", \"#D62728\", \"#9467BD\", \"#2CA02C\", \"#FF7F0E\",\n",
    "                \"#6B8E23\", \"#C71585\", \"#20B2AA\", \"#FFD700\"]\n",
    "    plt.figure(figsize=(15,5))\n",
    "    barWidth = 0.15\n",
    "    xAxis = np.arange(len(index))\n",
    "    offset1 = xAxis-barWidth\n",
    "    offset2 = xAxis\n",
    "    offset3 = xAxis+barWidth\n",
    "\n",
    "    # Neutral\n",
    "    plt.bar(offset2, a/total, barWidth, label = labels[0], color = colors[0])\n",
    "    bar2 = a/total\n",
    "    plt.bar(offset2, b/total, barWidth, label = labels[11], bottom = bar2, color = colors[1])\n",
    "    bar2 += b/total\n",
    "    plt.bar(offset2, c/total, barWidth, label = labels[17], bottom = bar2, color = colors[2])\n",
    "    bar2 += c/total\n",
    "    \n",
    "    # Benefitial \n",
    "    plt.bar(offset1, ab/total, barWidth, label = labels[5], color = colors[3])\n",
    "    bar1 = ab/total\n",
    "    plt.bar(offset1, abc/total, barWidth, label = labels[7], bottom = bar1, color = colors[4])\n",
    "    bar1 += abc/total\n",
    "    plt.bar(offset1, bc/total, barWidth, label = labels[15], bottom = bar1, color = colors[5])\n",
    "    bar1 += bc/total\n",
    "\n",
    "    # Detrimental\n",
    "    plt.bar(offset3, aa/total, barWidth, label = labels[1], color = colors[6])\n",
    "    bar3 = aa/total\n",
    "    plt.bar(offset3, aaa/total, barWidth, label = labels[2], bottom = bar3, color = colors[7])\n",
    "    bar3 += aaa/total\n",
    "    plt.bar(offset3, aab/total, barWidth, label = labels[3], bottom = bar3, color = colors[8])\n",
    "    bar3 += aab/total\n",
    "    plt.bar(offset3, aac/total, barWidth, label = labels[4], bottom = bar3, color = colors[9])\n",
    "    bar3 += aac/total\n",
    "    plt.bar(offset3, abb/total, barWidth, label = labels[6], bottom = bar3, color = colors[10])\n",
    "    bar3 += abb/total\n",
    "    plt.bar(offset3, ac/total, barWidth, label = labels[8], bottom = bar3, color = colors[11])\n",
    "    bar3 += ac/total\n",
    "    plt.bar(offset3, acb/total, barWidth, label = labels[9], bottom = bar3, color = colors[12])\n",
    "    bar3 += acb/total\n",
    "    plt.bar(offset3, acc/total, barWidth, label = labels[10], bottom = bar3, color = colors[13])\n",
    "    bar3 += acc/total\n",
    "    plt.bar(offset3, bb/total, barWidth, label = labels[12], bottom = bar3, color = colors[14])\n",
    "    bar3 += bb/total\n",
    "    plt.bar(offset3, bbb/total, barWidth, label = labels[13], bottom = bar3, color = colors[15])\n",
    "    bar3 += bbb/total\n",
    "    plt.bar(offset3, bbc/total, barWidth, label = labels[14], bottom = bar3, color = colors[16])\n",
    "    bar3 += bbc/total\n",
    "    plt.bar(offset3, bcc/total, barWidth, label = labels[16], bottom = bar3, color = colors[17])\n",
    "    bar3 += bcc/total\n",
    "    plt.bar(offset3, cc/total, barWidth, label = labels[18], bottom = bar3, color = colors[18])\n",
    "    bar3 += cc/total\n",
    "    plt.bar(offset3, ccc/total, barWidth, label = labels[19], bottom = bar3, color = colors[19])\n",
    "    bar3 += ccc/total\n",
    "\n",
    "    plt.legend(loc='right', bbox_to_anchor=(1.2,0.5))\n",
    "    plt.title(f\"{titleName}\")\n",
    "    if len(index) <= 300:\n",
    "        plt.xticks(xAxis, index, rotation = -90)\n",
    "    plt.ylabel('Probabilty')\n",
    "    plt.xlabel('Sequence Index')\n",
    "    plt.ylim(0,1)\n",
    "    return total\n",
    "\n",
    "def rpaConcentrations(primerConcentration, blockerConcentration, blockers):\n",
    "    ''' rpaConcentrations graphs 3 strand data particularly for RPA \n",
    "    Parameters:     primerConcentration     Concentration for the primers\n",
    "                    blockerConcentration    Concentration for the blockers\n",
    "                    blockers                List of blocker sequences (DNA)\n",
    "    Returns:        concList                List of concentrations after reaction\n",
    "    '''\n",
    "    a, aa, aaa, aab, aac, ab, abb, abc, ac, acb, acc = [], [],[],[],[], [], [],[],[],[],[]\n",
    "    b, bb, bbb, bbc, bc, bcc = [],[],[],[],[],[]\n",
    "    c, cc, ccc, total = [], [], [], []\n",
    "    concList = [a, aa, aaa, aab, aac, ab, abb, abc, ac, acb, acc,b, bb, bbb, bbc, bc, bcc, c, cc, ccc]\n",
    "\n",
    "    for i in range(len(blockers)):\n",
    "        A = Strand(fwdPrimer, material=nucleicAcid, name='A') # forward primer\n",
    "        B = Strand(blockers[i], material=nucleicAcid, name='B') #Blocker\n",
    "        C = Strand(rvsPrimer, material=nucleicAcid, name='C') # reverse primer\n",
    "        \n",
    "        # Specify complexes up to size 3\n",
    "        t1 = Tube(strands={A: primerConcentration, B: blockerConcentration, C: primerConcentration},\n",
    "                    complexes=SetSpec(max_size=3), name='t1')\n",
    "        my_results = tube_analysis(tubes=[t1], model=my_model)\n",
    "        complex_results = complex_analysis(complexes=t1,model=my_model, compute=['pfunc'])\n",
    "        complex_results2 = complex_concentrations(tube=t1,data=complex_results)\n",
    "        [concList[i].append(conc) for i, conc in enumerate(complex_results2.tubes[t1].complex_concentrations.values())]\n",
    "        total.append(sum(complex_results2.tubes[t1].complex_concentrations.values()))\n",
    "    return concList"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Accessing previously saved sequences \n",
    "if openExistingFile_Boolean == True:\n",
    "    csv.field_size_limit(sys.maxsize)\n",
    "\n",
    "    # Lists\n",
    "    index_list = []\n",
    "    gap = []\n",
    "    probe_sequences = []\n",
    "\n",
    "    # CSV file\n",
    "    with open(openExistingFile_Name, mode='r', newline='') as csv_file:\n",
    "        csv_reader = csv.DictReader(csv_file)\n",
    "\n",
    "        for row in csv_reader:\n",
    "            index_list.append(row['Index'])\n",
    "    \n",
    "            forward_gap_sequence = ast.literal_eval(row['Forward Gap Sequence 3-5'])[0]\n",
    "            gap.append(forward_gap_sequence)\n",
    "            \n",
    "            probe_sequence = ast.literal_eval(row['Probe Sequence 5-3'])\n",
    "            probe_sequences.append(probe_sequence)\n",
    "\n",
    "    probeLen, gapLen = [], []\n",
    "    probeIndex = range(n)\n",
    "    cProbe = probe_sequences[0]\n",
    "    for probe in cProbe:\n",
    "        probeLen.append(len(probe))\n",
    "\n",
    "    for seq in gap:\n",
    "        gapLen.append(len(seq))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Generating probe and gap sequences\n",
    "# Main\n",
    "if saveProbeGap_Sequences == True:\n",
    "    cProbe, probeLen, probeIndex = [], [], []\n",
    "    gap, gapLen = [], []\n",
    "    generateSequences(nitrogenousBase, gcMin, gcMax, gapLenMin, gapLenMax, \"Gap\", 1)\n",
    "    generateSequences(nitrogenousBase, gcMin, gcMax, probeLenMin, probeLenMax, \"Probe\", n)\n",
    "    probeIndex = range(n)\n",
    "\n",
    "    # Printing to csv file\n",
    "    csvFile = open(saveSequences_Name, 'wt', newline='')\n",
    "    writer = csv.writer(csvFile)    \n",
    "    try:\n",
    "        writer = csv.writer(csvFile)\n",
    "        writer.writerow(('Index', 'Forward Gap Sequence 3-5', 'Probe Sequence 5-3'))\n",
    "        for i in range(len(gap)):\n",
    "            if i == 0:\n",
    "                writer.writerow((i, gap, cProbe))\n",
    "            else:\n",
    "                writer.writerow((i, None, cProbe))\n",
    "    finally:\n",
    "        csvFile.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Processing Aptamer and Blocker strands\n",
    "# Lists\n",
    "strandA, strandAA, strandB, strandBB, strandAB = [], [], [], [], [] # strands\n",
    "strandALen, strandBLen = [], []\n",
    "gibbsEnergyA, gibbsEnergyAA, gibbsEnergyB, gibbsEnergyBB, gibbsEnergyAB = [], [], [], [], [] # gibbs free energy\n",
    "mfeA, mfeAA, mfeB, mfeBB, mfeAB = [], [], [], [], [] # minimal free energy\n",
    "concA, concB, concAA, concAB, concBB, totalConc = [], [], [], [], [], [] # concentration\n",
    "concList = [concA, concAA, concAB, concB, concBB]\n",
    "\n",
    "# Collecting Data\n",
    "for i in range(len(probeIndex)):\n",
    "    blockerSeq = cFwdPrimer+gap[0]+cDNAorRNA(cProbe[i])+cRvsPrimer+sBinding\n",
    "    aptamerSeq = cProbe[i]+rvsPrimer+BindingSite+Segment2\n",
    "    strandALen.append(len(aptamerSeq))\n",
    "    strandBLen.append(len(blockerSeq))\n",
    "    \n",
    "    A = Strand(aptamerSeq, material=nucleicAcid, name = 'A') # aptamer sequence\n",
    "    B = Strand(blockerSeq, material=nucleicAcid, name = 'B') # blocker sequence\n",
    "    t1 = Tube(strands= {A: concentration, B: concentration}, complexes=SetSpec(max_size=sizeMax), name='t1')\n",
    "    my_results = tube_analysis(tubes=[t1], model= my_model, compute=['mfe'])\n",
    "    \n",
    "    packagingData(my_results['(A)'], aptamerSeq, strandA, gibbsEnergyA, mfeA) # Collecting Strand A data\n",
    "    packagingData(my_results['(A+A)'], None, strandAA, gibbsEnergyAA, mfeAA) # Collecting Strand A-A data\n",
    "    packagingData(my_results['(B)'], blockerSeq, strandB, gibbsEnergyB, mfeB) # Collecting Strand B data\n",
    "    packagingData(my_results['(B+B)'], None, strandBB, gibbsEnergyBB, mfeBB) # Collecting Strand B-B data\n",
    "    packagingData(my_results['(A+B)'], None, strandAB, gibbsEnergyAB, mfeAB) # Collecting Strand A-B data\n",
    "\n",
    "    complex_results = complex_analysis(complexes=t1,model=my_model, compute=['pfunc'])\n",
    "    complex_results2 = complex_concentrations(tube=t1,data=complex_results)\n",
    "    [concList[i].append(conc) for i, conc in enumerate(complex_results2.tubes[t1].complex_concentrations.values())]\n",
    "    totalConc.append(sum(complex_results2.tubes[t1].complex_concentrations.values()))\n",
    "\n",
    "    if i%250 == 0 and i != 0:\n",
    "        print(i)\n",
    "\n",
    "# Combining data for analysis\n",
    "combinedData = np.column_stack((strandA, strandB,\n",
    "                                gibbsEnergyA, gibbsEnergyB, gibbsEnergyAA, gibbsEnergyAB, gibbsEnergyBB,\n",
    "                                mfeA, mfeB, mfeAA, mfeAB, mfeBB,\n",
    "                                probeIndex, strandALen, strandBLen,\n",
    "                                concA, concB, concAA, concAB, concBB, totalConc))\n",
    "\n",
    "# Changing matrix to dataframe\n",
    "labels = ['Aptamer (A)','Blocker (B)',\n",
    "          'Gibbs Energy of A','Gibbs Energy of B', 'Gibbs Energy of AA','Gibbs Energy of AB','Gibbs Energy of BB',\n",
    "          'MFE A', 'MFE B', 'MFE AA', 'MFE AB', 'MFE BB',\n",
    "          'Index', 'Length of A', 'Length of B',\n",
    "          'Concentration A', 'Concentration B', 'Concentration AA', 'Concentration AB', 'Concentration BB', 'Total Concentration']\n",
    "labeledCombinedData = pd.DataFrame(combinedData, columns=labels)\n",
    "labeledCombinedData.to_excel(saveData_PreFilters_Name)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Filtering by gibbs energy \n",
    "# List\n",
    "filteredGibbsData = []\n",
    "\n",
    "# Main\n",
    "for i in range(len(probeIndex)):\n",
    "    aptamerA = labeledCombinedData.iloc[i,0]\n",
    "    blockers = labeledCombinedData.iloc[i,1]\n",
    "    mfeAData = labeledCombinedData.iloc[i,7]\n",
    "    mfeBData = labeledCombinedData.iloc[i,8]\n",
    "\n",
    "    if gPattern.search(aptamerA[:5]) or gPattern.search(blockers[:5]):\n",
    "        continue\n",
    "    if consecutiveBP.search(aptamerA) or consecutiveBP.search(blockers):\n",
    "        continue\n",
    "\n",
    "    if minGibbs <= float(mfeAData) <= maxGibbs and minGibbs <= float(mfeBData) <= maxGibbs:\n",
    "        rowData = labeledCombinedData.iloc[i].tolist()\n",
    "        filteredGibbsData.append(rowData)\n",
    "\n",
    "labeledFilteredGibbsData = pd.DataFrame(filteredGibbsData, columns=labeledCombinedData.columns)\n",
    "filteredGibbsData = np.column_stack(labeledFilteredGibbsData.values.T)\n",
    "print(f\"Number of Sequence Pairs: {len(labeledFilteredGibbsData)}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Filtering by Aptamer-Blocker concentration\n",
    "# Main\n",
    "dataAB = labeledFilteredGibbsData.iloc[:,18].astype(float)\n",
    "dataTotal = labeledFilteredGibbsData.iloc[:,20].astype(float)\n",
    "concFilter = (dataAB/dataTotal) > 0.5\n",
    "\n",
    "filteredConcData = np.delete(labeledFilteredGibbsData.to_numpy(), np.where(~concFilter), axis=0)\n",
    "labeledFilteredConcData = pd.DataFrame(filteredConcData, columns=labeledFilteredGibbsData.columns)\n",
    "print(f\"Number of Sequence Pairs: {len(labeledFilteredConcData)}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Graphing and processing the Aptamer, Blocker, and CAP sequences\n",
    "# Lists\n",
    "a, aa, aaa, aab, aac, ab, abb, abc, ac, acb, acc = [], [],[],[],[], [], [],[],[],[],[]\n",
    "b, bb, bbb, bbc, bc, bcc = [],[],[],[],[],[]\n",
    "c, cc, ccc, totalABCapConc = [], [], [], []\n",
    "concList = [a, aa, aaa, aab, aac, ab, abb, abc, ac, acb, acc,b, bb, bbb, bbc, bc, bcc, c, cc, ccc]\n",
    "\n",
    "# Constants\n",
    "capstoneData = filteredConcData\n",
    "capstoneIndex = [row[12] for row in filteredConcData]\n",
    "barWidth = 0.15\n",
    "\n",
    "# Main\n",
    "for i in range(len(capstoneData)):\n",
    "    A = Strand(capstoneData[i,0], material=nucleicAcid, name='A') #Aptamer\n",
    "    B = Strand(capstoneData[i,1], material=nucleicAcid, name='B') #Blocker\n",
    "    C = Strand(cDNAorRNA(BindingSite+Segment2), material=nucleicAcid, name='C') #CAP \n",
    "\n",
    "    # Specify complexes up to size 3\n",
    "    t1 = Tube(strands={A: concentration, B: concentration, C: concentration},\n",
    "                complexes=SetSpec(max_size=3), name='t1')\n",
    "    my_results = tube_analysis(tubes=[t1], model=my_model)\n",
    "    complex_results = complex_analysis(complexes=t1,model=my_model, compute=['pfunc'])\n",
    "    complex_results2 = complex_concentrations(tube=t1,data=complex_results)\n",
    "    [concList[i].append(conc) for i, conc in enumerate(complex_results2.tubes[t1].complex_concentrations.values())]\n",
    "    totalABCapConc.append(sum(complex_results2.tubes[t1].complex_concentrations.values()))\n",
    "\n",
    "    if i%250 == 0 and i != 0:\n",
    "        print(i)\n",
    "\n",
    "testABC = np.column_stack((capstoneData[:,0], capstoneData[:,1], \n",
    "                            a, aa, aaa, aab, aac, ab, abb, abc, ac, acb, acc,\n",
    "                            b, bb, bbb, bbc, bc, bcc, c, cc, ccc, \n",
    "                            totalABCapConc, capstoneIndex))\n",
    "\n",
    "# Changing matrix to dataframe\n",
    "labels =['Aptamer (A)','Blocker (B)',\n",
    "         'A', 'AA', 'AAA', 'AAB', 'AAC', 'AB', 'ABB', 'ABC', 'AC', 'ACB', 'ACC',\n",
    "         'B', 'BB', 'BBB', 'BBC', 'BC', 'BCC', 'C', 'CC', 'CCC',\n",
    "         'Total Concentration', 'Index']\n",
    "labeledCombinedDataABCTesting = pd.DataFrame(testABC, columns=labels)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Filtering by Aptamer-CAP concentration\n",
    "# Main\n",
    "dataAB = labeledCombinedDataABCTesting.iloc[:,7].astype(float)\n",
    "dataAC = labeledCombinedDataABCTesting.iloc[:,10].astype(float)\n",
    "dataTotal = labeledCombinedDataABCTesting.iloc[:,22].astype(float)\n",
    "abcFilter = (dataAC/dataTotal) > (dataAB/dataTotal)\n",
    "\n",
    "filteredABCConcData = np.delete(labeledCombinedDataABCTesting.to_numpy(), np.where(~abcFilter), axis=0)\n",
    "labeledFilteredABCConcData = pd.DataFrame(filteredABCConcData, columns=labeledCombinedDataABCTesting.columns)\n",
    "print(f\"Number of Sequence Pairs: {len(labeledFilteredABCConcData)}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Reframing the concentration graph for the remaining sequences\n",
    "# Main\n",
    "totalConc2, concA2, concB2, concAA2, concAB2, concBB2 = [], [], [], [], [], [] # concentration\n",
    "concList2 = [concA2, concAA2, concAB2, concB2, concBB2]\n",
    "\n",
    "for i in range(len(labeledFilteredABCConcData)):\n",
    "    A = Strand(labeledFilteredABCConcData.iloc[i,0], material=nucleicAcid, name='A') #Aptamer\n",
    "    B = Strand(labeledFilteredABCConcData.iloc[i,1], material=nucleicAcid, name='B') #Blocker\n",
    "    \n",
    "    # Specify complexes up to size 3\n",
    "    t1 = Tube(strands={A: concentration, B: concentration},\n",
    "              complexes=SetSpec(max_size=2), name='t1')\n",
    "    my_results = tube_analysis(tubes=[t1], model=my_model)\n",
    "    \n",
    "    complex_results = complex_analysis(complexes=t1,model=my_model, compute=['pfunc'])\n",
    "    complex_results2 = complex_concentrations(tube=t1,data=complex_results)\n",
    "    [concList2[i].append(conc) for i, conc in enumerate(complex_results2.tubes[t1].complex_concentrations.values())]\n",
    "    totalConc2.append(sum(complex_results2.tubes[t1].complex_concentrations.values()))\n",
    "\n",
    "    if i%250 == 0 and i != 0:\n",
    "        print(i)\n",
    "\n",
    "testing2 = np.column_stack((labeledFilteredABCConcData.iloc[:,0], labeledFilteredABCConcData.iloc[:,1],\n",
    "                                concA2, concAA2, concAB2,\n",
    "                                concB2, concBB2, \n",
    "                                totalConc2, labeledFilteredABCConcData.iloc[:,23]))\n",
    "\n",
    "# Changing matrix to dataframe\n",
    "labels = ['Aptamer (A)','Blocker (B)',\n",
    "          'Equilibrium Concentration A', 'Equilibrium Concentration AA', 'Equilibrium Concentration AB', \n",
    "          'Equilibrium Concentration B','Equilibrium Concentration BB', \n",
    "          'Total Concentration','Index']\n",
    "labeledCombinedDataTesting2 = pd.DataFrame(testing2, columns=labels) # Converting matrix into dataframe to add labels"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Graphing the concentrations of the remaining sequences\n",
    "# Main\n",
    "concA2, concAA2, concAB2, concB2, concBB2 = [np.array(data) for data in concList2]\n",
    "capstoneIndex = labeledCombinedDataTesting2.iloc[:,8]\n",
    "plt.figure(7, figsize=(15,15))\n",
    "plt.subplot(2,1,1)\n",
    "barWidth = 0.15\n",
    "xAxis = np.arange(len(capstoneIndex))\n",
    "plt.bar(xAxis, (concA2/totalConc2), barWidth, label = \"Aptamer (A)\")\n",
    "bottomValue1 = np.array(concA2/totalConc2)\n",
    "plt.bar(xAxis, concB2/totalConc2, barWidth, label=\"Blocker (B)\", bottom = bottomValue1)\n",
    "bottomValue1 += np.array(concB2/totalConc2)\n",
    "plt.bar(xAxis+barWidth, concAA2/totalConc2, barWidth, label=\"AA\")\n",
    "bottomValue2 = np.array(concAA2/totalConc2)\n",
    "plt.bar(xAxis+barWidth, concAB2/totalConc2, barWidth, label=\"AB\", bottom = bottomValue2)\n",
    "bottomValue2 += np.array(concAB2/totalConc2)\n",
    "plt.bar(xAxis+barWidth, concBB2/totalConc2, barWidth, label=\"BB\", bottom = bottomValue2)\n",
    "bottomValue2 += np.array(concBB2/totalConc2)\n",
    "plt.legend(loc='right', bbox_to_anchor=(1.15,0.9))\n",
    "plt.title('Complex Probabilities')\n",
    "plt.xticks(xAxis+(1.5)*barWidth, capstoneIndex, rotation = -90)\n",
    "plt.ylabel('Probabilty')\n",
    "plt.xlabel('Index')\n",
    "plt.ylim(0,1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Second Filter for Aptamer-CAP concentrations\n",
    "# Main\n",
    "dataAB = labeledFilteredABCConcData.iloc[:,7].astype(float)\n",
    "dataAC = labeledFilteredABCConcData.iloc[:,10].astype(float)\n",
    "dataB = labeledFilteredABCConcData.iloc[:,13].astype(float)\n",
    "dataTotal = labeledFilteredABCConcData.iloc[:,22].astype(float)\n",
    "abcFilter2 = (dataAC+dataB)/dataTotal > 0.2\n",
    "\n",
    "# Filtering ABC array\n",
    "secondFilteredABCConcData = np.delete(labeledFilteredABCConcData.to_numpy(), np.where(~abcFilter2), axis=0)\n",
    "labeledSecondFilteredABCConcData = pd.DataFrame(secondFilteredABCConcData, columns=labeledFilteredABCConcData.columns)\n",
    "\n",
    "# Filtering AB array\n",
    "secondFilteredTesting2 = np.delete(labeledCombinedDataTesting2.to_numpy(), np.where(~abcFilter2), axis=0)\n",
    "labeledSecondFilteredTesting2 = pd.DataFrame(secondFilteredTesting2, columns=labeledCombinedDataTesting2.columns)\n",
    "if len(labeledSecondFilteredABCConcData) == len(labeledSecondFilteredTesting2):\n",
    "    print(\"AB and ABC Pairs Match!\")\n",
    "    print(f\"Number of Sequence Pairs: {len(labeledSecondFilteredABCConcData)}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Filtering and Graphing Final Sequences\n",
    "# Main\n",
    "abColumns = list(range(2,7))\n",
    "abcColumns = list(range(2, 22))\n",
    "\n",
    "finalABConc = [labeledSecondFilteredTesting2.iloc[:,i].astype(float) for i in abColumns]\n",
    "concA3, concAA3, concAB3, concB3, concBB3 = [np.array(data) for data in finalABConc]\n",
    "totalAB = np.sum(finalABConc, axis=0)\n",
    "capstoneIndex = labeledSecondFilteredTesting2.iloc[:,8]\n",
    "\n",
    "finalABCConc = [labeledSecondFilteredABCConcData.iloc[:,i].astype(float) for i in abcColumns]\n",
    "finalIndex = labeledSecondFilteredABCConcData.iloc[:,23]\n",
    "\n",
    "# Graphing AB\n",
    "plt.figure(7, figsize=(15,15))\n",
    "plt.subplot(2,1,1)\n",
    "barWidth = 0.15\n",
    "xAxis = np.arange(len(capstoneIndex))\n",
    "plt.bar(xAxis, (concA3/totalAB), barWidth, label = \"Aptamer (A)\")\n",
    "bottomValue1 = np.array(concA3/totalAB)\n",
    "plt.bar(xAxis, concB3/totalAB, barWidth, label=\"Blocker (B)\", bottom = bottomValue1)\n",
    "bottomValue1 += np.array(concB3/totalAB)\n",
    "plt.bar(xAxis+barWidth, concAA3/totalAB, barWidth, label=\"AA\")\n",
    "bottomValue2 = np.array(concAA3/totalAB)\n",
    "plt.bar(xAxis+barWidth, concAB3/totalAB, barWidth, label=\"AB\", bottom = bottomValue2)\n",
    "bottomValue2 += np.array(concAB3/totalAB)\n",
    "plt.bar(xAxis+barWidth, concBB3/totalAB, barWidth, label=\"BB\", bottom = bottomValue2)\n",
    "bottomValue2 += np.array(concBB3/totalAB)\n",
    "plt.legend(loc='right', bbox_to_anchor=(1.15,0.9))\n",
    "plt.title('Complex Probabilities')\n",
    "plt.xticks(xAxis+(1.5)*barWidth, capstoneIndex, rotation = -90)\n",
    "plt.ylabel('Probabilty')\n",
    "plt.xlabel('Index')\n",
    "plt.ylim(0,1)\n",
    "\n",
    "# Graphing ABC\n",
    "graphingLabels = ['Aptamer (A)', 'AA', 'AAA', 'AAB', 'AAC', 'AB', 'ABB', 'ABC', 'AC', 'ACB', 'ACC',\n",
    "                  'Blocker (B)', 'BB', 'BBB', 'BBC', 'BC', 'BCC',\n",
    "                  'CAP (C)', 'CC', 'CCC',]\n",
    "finalTotals = graphingAptamerBlockerCAP(finalABCConc, finalIndex, \"Final Complex Probabilities\", graphingLabels)\n",
    "labeledSecondFilteredABCConcData.to_excel(saveData_PostFilters_Name) "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Graphing the concentrations of the forward primer, reverse primer, and blocker sequence \n",
    "\n",
    "# THIS IS ONLY FOR THERMODYNAMIC AMPLIFICATION (not enzamatic therefore not applicable to RPA)\n",
    "\n",
    "# Main\n",
    "finalAptamers = labeledSecondFilteredABCConcData.iloc[:,0]\n",
    "blockerRPA = labeledSecondFilteredABCConcData.iloc[:,1]\n",
    "indexRPA = labeledSecondFilteredABCConcData.iloc[:,23]\n",
    "\n",
    "concList1To1 = rpaConcentrations(5e-7, 5e-7, blockerRPA)\n",
    "concList2To1 = rpaConcentrations(1e-6, 5e-7, blockerRPA)\n",
    "concList1To2 = rpaConcentrations(5e-7, 1e-6, blockerRPA)\n",
    "\n",
    "graphingRPALabels = ['Foward Primer (F)', 'FF', 'FFF', 'FFB', 'FFR', 'FB', 'FBB', 'FBR', 'FR', 'FRB', 'FRR',\n",
    "                  'Blocker (B)', 'BB', 'BBB', 'BBR', 'BR', 'BRR',\n",
    "                  'Reverse Primer (R)', 'RR', 'RRR',]\n",
    "finalTotals = graphingConcentration3Strand(concList1To1, indexRPA, \"1:1 Ratio Complex Probability with Blocker and Primer\", graphingRPALabels)\n",
    "finalTotals = graphingConcentration3Strand(concList2To1, indexRPA, \"2:1 Ratio Complex Probability with Blocker and Primer\", graphingRPALabels)\n",
    "finalTotals = graphingConcentration3Strand(concList1To2, indexRPA, \"1:2 Ratio Complex Probability with Blocker and Primer\", graphingRPALabels)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Printing the blockers and aptamer sequences\n",
    "print(\"Blockers\")\n",
    "for data in blockerRPA:\n",
    "    print(data)\n",
    "\n",
    "print(\"Aptamers\")\n",
    "for data in finalAptamers:\n",
    "    print(data)\n",
    "\n",
    "print(f\"Gap: {gap[0]}\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "base",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
